File: leetcode_75/Graphs_DFS/python/841.py
from typing import List


class Solution:

    def dfs(self, node, graph, visited):
        visited[node] = True
        for v in graph[node]:
            if not visited[v]:
                self.dfs(v, graph, visited)

    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        graph = dict()
        for i in range(len(rooms)):
            graph[i] = rooms[i]

        visited = [False] * len(graph)
        visited[0] = True
        root = 0
        self.dfs(root, graph, visited)

        return True if all(x for x in visited) else False

File: leetcode_75/Graphs_DFS/cpp/841.cpp
#include <vector>
#include <unordered_map>

using namespace std;

class Solution
{
public:
    bool canVisitAllRooms(vector<vector<int>> &rooms)
    {
        unordered_map<int, vector<int>> graph;
        int n = rooms.size();

        for (int i = 0; i < n; ++i)
        {
            graph[i] = rooms[i];
        }

        vector<bool> visited(n, false);
        int root = 0;
        this->dfs(root, graph, visited);

        for (auto x : visited)
        {
            if (!x)
            {
                return false;
            }
        }
        return true;
    }

private:
    void dfs(int node, unordered_map<int, vector<int>> &graph, vector<bool> &visited)
    {
        visited[node] = true;

        for (auto v : graph[node])
        {
            if (!visited[v])
            {
                this->dfs(v, graph, visited);
            }
        }
    }
};